### ISSUE TITLE
[Bug]: numerical issues with cudagraph replay

### ISSUE THREAD

**Issue by AmdSampsa** [ROCmSoftwarePlatform/rocBLAS#1649]
## Summary

We have numerical issues when replaying models with cudagraph.

Unfortunately this bug is devilishly difficult to pinpoint exactly in the stack `pytorch -> miopen -> rocblas`, but key indicators suggest its about rocblas.

The problem manifests itself mostly with:

- deep convolutional models with lots of layers
- the `bfloat16` datatype
- when miopen goes into the conv_gemm backend -> **the only conv backend that uses rocblas**

I'm using these env variables when running pytorch:
```bash
export MIOPEN_DEBUG_CONV_FFT=0 # PASS
export MIOPEN_DEBUG_CONV_DIRECT=0 # PASS
export MIOPEN_DEBUG_CONV_GEMM=1 # FAIL
export MIOPEN_DEBUG_CONV_WINOGRAD=0 # PASS
export MIOPEN_DEBUG_CONV_IMPLICIT_GEMM=0 # PASS
# 
export MIOPEN_ENABLE_LOGGING=1
export MIOPEN_LOG_LEVEL=4
```
-> in the logs we can see that `rocblas_dgemm` (a level-3 algo) is heavily used.

On the other hand, [in the docs](https://rocm.docs.amd.com/projects/rocBLAS/en/latest/reference/beta-features.html#graph-support-for-rocblas) its said that hipgraphs are beta feature.

I've also enabled these ones:
```bash
export ROCBLAS_DEVICE_MEMORY_SIZE=1073741824
export ROCBLAS_STREAM_ORDER_ALLOC=1
```
and checked that the program runs in HOST pointer mode.  Btw, this bit in the docs is a bit confusing:

*"BLAS Level-3 and BLAS-EX functions in pointer mode device do not support HIP Graph"* -  on the other hand it says graphs need stream-ordered/async mem allocs (which we do not have in HOST pointer mode since its all sync mem alloc..?)

In any case, I have also tested a scenario where [reserving memory for miopen](https://github.com/pytorch/pytorch/blob/6abe450a6fda580714a276ed2dd695ab99074b93/aten/src/ATen/native/miopen/Conv_miopen.cpp#L301) has been fixed by reserving it statically before any graph capture takes place (among other quick/test-fixes).  Didn't help.

[Here](https://gist.github.com/AmdSampsa/bd58a18980fc5c31002ece0ddcdd1ac5) is the simplest repro I have been able to create on this matter.  It shows clearly how results differ when comparing normal vs. cudagraphs run.

Any attempt to run a small toy model, say a single convolution + batchnorm will not reveal the bug.

You can also try models at [pytorch/benchmarks/dynamo](https://github.com/pytorch/pytorch/tree/main/benchmarks/dynamo), say:

```bash
./timm_models.py --inference --accuracy --bfloat16 --backend=inductor --only dpn107
./timm_models.py --inference --accuracy --bfloat16 --backend=inductor --only pnaslarge05
```
which both fail (however, those scripts are very messy so its probably best to concentrate on the mentioned minimal repro).

Any help appreciated.

P. S. I'm thinking it would be a nice-to-have feature to have a pure c++ unit test (part of rocblas) that would test a bigger model with cudagaph rec and playback.

### Related issues

- https://github.com/ROCm/rocBLAS/issues/1240

### Environment

```bash
rocminfo | grep Marketing
  Marketing Name:          Intel(R) Xeon(R) Platinum 8468
  Marketing Name:          Intel(R) Xeon(R) Platinum 8468
  Marketing Name:          AMD Instinct MI300X
  Marketing Name:          AMD Instinct MI300X
  Marketing Name:          AMD Instinct MI300X
  Marketing Name:          AMD Instinct MI300X
  Marketing Name:          AMD Instinct MI300X
  Marketing Name:          AMD Instinct MI300X
  Marketing Name:          AMD Instinct MI300X
  Marketing Name:          AMD Instinct MI300X
```
```bash
dpkg -s rocblas | grep Version
Version: 4.4.0.60400-47~22.04
dpkg -s rocm-core | grep Version
Version: 6.4.0.60400-47~22.04
```

[environment.txt](https://github.com/user-attachments/files/20725239/environment.txt)

**Comment by ammallya** [comment]
Imported to ROCm/rocm-libraries
